miden-tx/src/host/tx_authenticator.rs

```diff
@@ -0,0 +1,170 @@
+use alloc::{collections::BTreeMap, string::ToString, vec::Vec};
+use core::cell::RefCell;
+
+use miden_objects::{
+    accounts::AccountDelta,
+    crypto::dsa::rpo_falcon512::{self, Polynomial},
+};
+use rand::Rng;
+use vm_processor::{Digest, Felt, Word};
+
+use crate::error::AuthenticationError;
+
+// TRANSACTION AUTHENTICATOR
+// ================================================================================================
+
+/// Defines an authenticator for transactions.
+///
+/// The main purpose of the authenticator is to generate signatures for a given message against
+/// a key managed by the authenticator. That is, the authenticator maintains a set of public-
+/// private key pairs, and can be requested to generate signatures against any of the managed keys.
+///
+/// The public keys are defined by [Digest]'s which are the hashes of the actual public keys.
+pub trait TransactionAuthenticator {
+    /// Retrieves a signataure for a specific message as a list of [Felt].
+    /// The request is initiaed by the VM as a consequence of the SigToStack advice
+    /// injector.
+    ///
+    /// - `pub_key`: The public key used for signature generation.
+    /// - `message`: The message to sign, usually a commitment to the transaction data.
+    /// - `account_delta`: An informational parameter describing the changes made to
+    ///   the account up to the point of calling `get_signature()`. This allows the
+    ///   authenticator to review any alterations to the account prior to signing.
+    ///   It should not be directly used in the signature computation.
+    fn get_signature(
+        &self,
+        pub_key: Word,
+        message: Word,
+        account_delta: &AccountDelta,
```

Comment:
Not for this PR, but I think in the future we should consider making this `account_delta: Option<&AccountDelta>` to cover cases when signatures may need to be generated for other purposes besides transaction authentication. But this requires more thinking.